[Serialization] support loading torch state dict from disk

hanouticelina opened the pull request:
Implement helpers to load a torch state dict from disk. For the implementation, it's mostly an importing from transformers and diffusers implementation with additional error handling and some refactoring. the loading can be done from a single file or from shards. It handles both safetensors and pickle files. Saving torch state dict has been previously added in #2314.

---
This PR:
- adds `load_torch_model()` helper function that takes a `nn.Module` and  a checkpoint path (either a single file or a directory) as input and loads the weights into the model.

usage example:
```python
from huggingface_hub import load_torch_model
model = ... # A PyTorch model

# load the weights into model
load_torch_model(model, "path/to/checkpoint")
```
- adds two low-level helpers that can be used directly by transformers, diffusers and accelerate:
  - `load_sharded_checkpoint()`: loads sharded weights from disk.
  - `load_state_dict_from_file()`: loads a single checkpoint file.
- tests have been added and documentation have also been updated.
---
Note: PRs will be opened in transformers, diffusers and accelerate to integrate these helpers once huggingface_hub v0.27.0 is released.

cc @sayakpaul for diffusers, @muellerzr and @SunMarc for accelerate and @ArthurZucker for transformers. Happy to get any feedback on this. The goal is the same as for the saving helpers: standardize things across our libraries and establish consistent conventions!

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_2687). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+        filename_pattern = (
+            constants.SAFETENSORS_WEIGHTS_FILE_PATTERN if safe else constants.PYTORCH_WEIGHTS_FILE_PATTERN
+        )
+        index_file = checkpoint_path / (filename_pattern.format(suffix="") + ".index.json")
```
Here if `safe=True`, it's clear that we want to load from `safetensors`. But if `safe=False`, we could still try to load from `safetensors` if it's installed and the weights are available in that format. So maybe we could have a logic like (pseudo-code):

```py
index_file = checkpoint_path / (constants.SAFETENSORS_WEIGHTS_FILE_PATTERN.format(suffix="") + ".index.json")
if not index_file.is_file() and safe:
    index_file := checkpoint_path / (constants.PYTORCH_WEIGHTS_FILE_PATTERN.format(suffix="") + ".index.json")
if index_file.is_file():
    ... # load

... # else, same as now: check if single file exists or raise error
```

hanouticelina replied:
The logic to prioritize loading safetensors makes sense since the [docstring](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/serialization/_torch.py#L97) for `safe_serialization` states that "Saving as pickle is deprecated and will be removed in a future version."
this is addressed in [0cce776](https://github.com/huggingface/huggingface_hub/pull/2687/commits/0cce776851f707a8ac9a7fc0fbfba4353882384b), I'd appreciate your opinion on this specific part of the code!

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+
+    elif checkpoint_path.is_dir():
+        # Check for index file which indicates sharded checkpoint
+        filename_pattern = (
+            constants.SAFETENSORS_WEIGHTS_FILE_PATTERN if safe else constants.PYTORCH_WEIGHTS_FILE_PATTERN
+        )
```
I think we should make `filename_pattern` an optional parameter as well (same as for saving). In diffusers for instance, some components can be saved both in `model.safetensors` and `model.fp16.safetensors` ([example](https://huggingface.co/stabilityai/sdxl-turbo/tree/main/text_encoder)). So they would need to somehow provide this information. For single file, it's fine by passing `checkpoints_path` pointing to the correct file but for sharded variants, they'll need to pass a pattern.

Wauplin replied:
or otherwise we could also support the case `load_torch_model(model, checkpoints_path="model.fp16.safetensors.index.json")` i.e. directly pass an index file

Wauplin replied:
(or maybe I'm overthinking it and no-one needs that level of granularity yet).

hanouticelina replied:
for sharded variants another example is https://huggingface.co/mistralai/Mistral-Large-Instruct-2411 in which we have two *.index.json (one for the weights compatible with mistral's format and one compatible with transformers) so yes it makes sense to add a `filename_pattern`as a parameter since the aim here is to be framework agnostic.

hanouticelina replied:
addressed in [0cce776](https://github.com/huggingface/huggingface_hub/pull/2687/commits/0cce776851f707a8ac9a7fc0fbfba4353882384b)

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+            "Expected either a sharded checkpoint with an index file, or a single model file."
+        )
+
+    else:
+        raise ValueError(f"Invalid checkpoint path: {checkpoint_path}")
```
I feel that we could get rid of `elif checkpoint_path.is_dir():` and these 2 lines. In practice we can assume that either `checkpoint_path`.is_file()` or `checkpoint_path.is_dir()` is True. Worse case, we can add a `if not checkpoint_path.exists(): raise ...` at the beginning of the method.

Removing the elif/else would deindent a large chunk of code which is better looking IMO

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+
+def load_torch_model(
+    model: "torch.nn.Module",
+    checkpoint_path: Union[str, os.PathLike],
```
(unrelated to this PR) I used to put `Union[str, Path]` in a lot of places in `huggingface_hub` but agree `Union[str, os.PathLike]` is more appropriate. Would be cool to harmonize this at some point (not urgent)

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -644,3 +979,18 @@ def _get_dtype_size(dtype: "torch.dtype") -> int:
     return _SIZE[dtype]
+
+
+class _IncompatibleKeys(namedtuple("IncompatibleKeys", ["missing_keys", "unexpected_keys"])):
+    """
+    This is used to report missing and unexpected keys in the state dict.
+    Taken from https://github.com/pytorch/pytorch/blob/main/torch/nn/modules/module.py#L52.
+
+    """
+
+    def __repr__(self) -> str:
+        if not self.missing_keys and not self.unexpected_keys:
+            return "<All keys matched successfully>"
+        return super().__repr__()
+
+    __str__ = __repr__
```
I would tend to prefer a dataclass here for consistency with the rest of the library

hanouticelina replied:
The choice of using a `NamedTuple` for `_IncompatibleKeys` is to maintain a consistency between the output of `_load_sharded_checkpoint()` and `nn.Module.load_state_dict()`. This makes it easier for the integration in transformers (I will have a PR for this shortly). I'd rather have this uniformity to simplify the handling of the return values of `load_torch_model()`.

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+    """
+    Loads a sharded checkpoint into a model. This is the same as
+    [`torch.nn.Module.load_state_dict`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=load_state_dict#torch.nn.Module.load_state_dict)
+    but for a sharded checkpoint. Each shard is loaded one by one and deleted after being loaded into the model.
```
> and deleted after being loaded into the model.

wrong copy-paste? or does it mean "removed from memory" ?

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+    # Check if file is empty (0 bytes)
+    if checkpoint_path.stat().st_size == 0:
+        raise OSError(
+            f"Checkpoint file at '{checkpoint_path}' exists but is empty. This might indicate a failed "
+            "download or a Git LFS file that hasn't been properly downloaded. If using Git LFS, please "
+            "ensure you have run 'git lfs pull' to download the actual model weights."
```
Lol, seems very legacy code ^^ Not opposed to keeping it though

hanouticelina replied:
I kept the same error handling as transformers but yes, this is not relevant anymore

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -349,6 +354,336 @@ def split_torch_state_dict_into_shards(
+        raise ValueError(f"Invalid checkpoint path: {checkpoint_path}")
+
+
+def load_sharded_checkpoint(
```
I wonder if we want to make this method private instead.

For what I see it does the same as `load_torch_model` but assumes the directory includes a sharded model while in `load_torch_model` there is a if/else logic depending on single-file/sharded. Is there a use case where a user would like to bypass this if/else logic?

hanouticelina replied:
indeed, there is no reason to bypass the if/else logic

hanouticelina replied:
done in [0cce776](https://github.com/huggingface/huggingface_hub/pull/2687/commits/0cce776851f707a8ac9a7fc0fbfba4353882384b)

Wauplin commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -357,7 +692,7 @@ def _get_unique_id(tensor: "torch.Tensor") -> Union[int, Tuple[Any, ...]]:
     try:
         # for torch 2.1 and above we can also handle tensor subclasses
-        from torch.utils._python_dispatch import is_traceable_wrapper_subclass
+        from torch.utils._python_dispatch import is_traceable_wrapper_subclass  # type: ignore[import]
```
Are all of these `# type: ignore[import]` mandatory? I don't have any issues locally with 

```sh
# pip freeze | grep -e torch -e mypy
mypy==1.5.1
mypy-extensions==1.0.0
pytorch-lightning==2.0.2
timm @ git+https://github.com/rwightman/pytorch-image-models.git@25ffac6880c2e9c441c070ffeff75d392b98f632
torch==2.5.1
torchaudio==2.3.1
torchmetrics==0.11.4
torchvision==0.20.1
```

hanouticelina replied:
I think my local linter config was somehow broken, it's fixed now

Wauplin reviewed (commented):
Nice PR @hanouticelina ! That's promising for our other libs :)  I've made a first pass on the PR and left a few comments. Overall looks good though I think we can update a few things to expose only 1 or 2 methods in the library. 

Also, would it be possible to open a PR on `transformers` to showcase how it would be used? No need to make updates everywhere in the lib', just 1 example is enough for now

hanouticelina commented:
@Wauplin thanks for the review! I've addressed the comments and created a draft PR [hanouticelina/transformers#1](https://github.com/hanouticelina/transformers/pull/1) to illustrate the integration (I've opened it on my personal fork for now while this PR is still WIP) –would love to have your feedback on that!

sayakpaul commented:
Sorry for the policing, but does a similar PR need to be opened in `diffusers` too? 👀

hanouticelina commented:
> Sorry for the policing, but does a similar PR need to be opened in `diffusers` too? 👀

@sayakpaul, yes! we plan to open a PR in `diffusers` once these helpers are released. For now, I've created [hanouticelina/transformers/pull/1](https://github.com/hanouticelina/transformers/pull/1) just as an example integration

Wauplin reviewed (approved):
Great job @hanouticelina ! This is a super well documented and tested PR which is much appreciated for such a key-part of the library! Tested it locally and it seems to work as I'd expect it^^

@LysandreJik @ArthurZucker would it be possible to take a closer look at this PR and especially https://github.com/hanouticelina/transformers/pull/1 to confirm everything's fine for you as well (worth case we ship and make hot-fixes if necessary). We'd like to ship this quickly :)

hanouticelina commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -371,6 +376,345 @@ def split_torch_state_dict_into_shards(
+        from torch import load
+    except ImportError as e:
+        raise ImportError(
+            "Please install `safetensors` to load safetensors checkpoint. "
+            "You can install it with `pip install safetensors`."
```
oops, an artifact from copy-pasting

hanouticelina commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -371,6 +376,345 @@ def split_torch_state_dict_into_shards(
+                f"The safetensors archive passed at {checkpoint_file} does not contain the valid metadata. Make sure "
+                "you save your model with the `save_torch_model` method."
+            )
+        return load_file(checkpoint_file)
```
indeed! thanks for pointing this out. 
Note that in transformers, we always load the state dict in cpu or in a meta device and then the weights get moved to their respective devices during [model dispatch](https://huggingface.co/docs/accelerate/en/package_reference/big_modeling#accelerate.dispatch_model) in the from_pretrained() method. 

Also meta device is not supported with `safetensors`. i guess we can fall back to cpu with a warning when meta device is specified in this case.

hanouticelina commented on src/huggingface_hub/serialization/_torch.py:
```diff
@@ -371,6 +376,345 @@ def split_torch_state_dict_into_shards(
+        map_location (`str` or `torch.device`, *optional*):
+            A `torch.device` object, string or a dict specifying how to remap storage locations. It
+            indicates the location where all tensors should be loaded.
+        weights_only (`bool`, *optional*, defaults to `False`):
+            If True, only loads the model weights without optimizer states and other metadata.
+            Only supported in PyTorch >= 1.13.
+        mmap (`bool`, *optional*, defaults to `False`):
+            Whether to use memory-mapped file loading. Memory mapping can improve loading performance
+            for large models in PyTorch >= 2.1.0 with zipfile-based checkpoints.
```
correct!

LysandreJik reviewed (approved):
From a quick look these two helpers could indeed replace a few codepaths in transformers. There is quite a bit of complexity there so I would recommend merging an initial version, setting up a branch in transformers to switch to these helpers and running the CI on it.

Thanks for your work @hanouticelina ! Any work that makes our `from_pretrained` method and our `modeling_utils` module simpler are welcome

hanouticelina commented:
thanks @LysandreJik, indeed, the best way to test these helpers is to run transformers CI on it! I will merge this and then will create the branch directly in transformers
